### PR TITLE
A new way to show status messages from esphome

### PR DESCRIPTION
**Pull request by nielsnl68**
# What does this implement/fix?

This is a fully customizing status report component that will use triggers to allow you to execute what you want to happen.
It gives you also the option to create custom status actions. that you can fire from within other triggers as well.

At the moment It exist of 4 types of triggers:

1. Eminent triggers. Those will direct show as soon as the are reported, For now those are the **Application Error** and **Application Warning** states. Before this triggers are fired all running status trigger will be stopped.
2. Network triggers. As long there is are network components setup the following triggers for **Network**, **api**, **mqtt** and **wifi_ap** connections are available.  
3. Custom triggers. You can create as many triggers as you need and fire them in any other trigger using `status.push: youtrigger` and remove them again using `status.pop: yourtrigger`
4. The clear trigger. This trigger will be fired when there is no eminent or network trigger configurated or when the custom trigger stack is empty. This is a required trigger.

For each custom trigger you can set a `group` name and `priority`.  The `group` name is used to pop all triggers with the same name.  Triggers with a lower `priority` will be put below triggers with a higher Priority in the view stack. And will only be fired until the trigger gets on top of the list.

When a trigger is pushed to the trigger stack all triggers with the same `group` or 'trigger_id' will be removed first before added to the list again.

This PR also adds a way to see of there is a network solution available and when Wifi used and AP is enabled you can see if the AP is setup or not.

the following triggers are available:
```yaml
   on_app_error  &&  on_clear_app_error
   on_app_warning  &&  on_clear_app_warning
   on_network_disconnected  &&  on_network_connected
   on_wifi_ap_enabled  &&  on_wifi_ap_disabled
   on_api_disconnected  && on_api_connected
   on_mgtt_disconnected  &&  on_mqtt_connected
   on_custom_status
   on_turn_off

```

I needed this to allow to use my rgb led for status messages.

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):** fixes <link to issue>

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#<esphome-docs PR number goes here>

## Test Environment

- [x] ESP32
- [ ] ESP32 IDF
- [ ] ESP8266
- [ ] RP2040
- [ ] BK72xx
- [ ] RTL87xx

## Example entry for `config.yaml`:

```yaml
# Example config.yaml
---
esphome:
  name: m5stack-atom-echo

esp32:
  board: m5stack-atom


light:
  - platform: esp32_rmt_led_strip
    id: led
    name: None
    disabled_by_default: true
    entity_category: config
    pin: GPIO27
    default_transition_length: 0s
    chipset: SK6812
    num_leds: 1
    rgb_order: grb
    rmt_channel: 0
    effects:
      - pulse:
          name: "Slow Pulse"
          transition_length: 250ms
          update_interval: 250ms
          min_brightness: 50%
          max_brightness: 100%
      - pulse:
          name: "Fast Pulse"
          transition_length: 100ms
          update_interval: 100ms
          min_brightness: 50%
          max_brightness: 100%


binary_sensor:
  - platform: gpio
    pin:
      number: GPIO39
      inverted: true
    name: Button
    disabled_by_default: true
    entity_category: diagnostic
    id: button
    on_press:
      - status.push:
          trigger_id: custom_status
    on_release:
      - status.pop:
          trigger_id: custom_status
    on_click:
    - min_length: 50ms
      max_length: 350ms
      then:
        - status.push:
            trigger_id: custom_status2
    - min_length: 500ms
      max_length: 1000ms
      then:
        - status.pop:
            trigger_id: custom_status2


status_indicator:
  on_clear_all: #Manditory
    - light.turn_off:
        id: led

  on_app_error:
    then:
    - light.turn_on:
        id: led
        red: 100%
        green: 0%
        blue: 0%
        brightness: 100%
        effect: Fast Pulse

  on_app_warning:
    then:
    - light.turn_on:
        id: led
        red: 100%
        green: 0%
        blue: 0%
        brightness: 100%
        effect: None

  on_custom_status:
    - trigger_id: custom_status
      then:
        - light.turn_on:
            id: led
            red: 0%
            green: 100%
            blue: 0%
            brightness: 100%
            effect: None
    - trigger_id: custom_status2
      then:
        - light.turn_on:
            id: led
            red: 0%
            green: 100%
            blue: 50%
            brightness: 100%
            effect: None

```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

**Comment by nagyrobi** [comment]
I like this because it guess it allows not only LEDs, but also printing messages on a display, if that part is still operational.

When do Application Error and Application Warning happen? Isn't ESPHome rebooting automatically on Application Error?

File: esphome/components/status_indicator/status_indicator.cpp [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+    if (status.empty() && not is_connected()) {
+      status = "on_network_disconnected";
+      this->status_.on_network = 1;
+    } else if (this->status_.on_network) {
+      status = "on_network_connected";
+      this->status_.on_network = 0;
+    }
+
+#ifdef USE_API
+    if (status.empty() && api::global_api_server != nullptr && not api::global_api_server->is_connected()) {
```
Comment (Unfocused) [inline esphome/components/status_indicator/status_indicator.cpp:93]: `on_api_connected` and `on_api_disconnected` here will apply to any API client, including the ESPHome CLI or the Home Assistant Addon connecting just for logging - with no way to tell what type of client connected/disconnected. At least with `api.on_client_disconnected` you can manually check the `client_info` string parameter starts with "Home Assistant".

That's a footgun for displaying an error/warning when something unimportant like a logging client disconnects but Home Assistant is still connected. And generally, Home Assistant will be the only client that matters in the context of a status indicator.

File: esphome/components/status_indicator/status_indicator.cpp [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+
+void StatusIndicator::push_trigger(StatusTrigger *trigger) {
+  this->pop_trigger(trigger, true);
+  ESP_LOGD(TAG, "Push ID: %s", trigger->get_info().c_str());
+
+  for (auto i = this->stack_.begin(); i != this->stack_.end(); ++i) {
+    StatusTrigger *st = *i;
+    if (trigger->get_priority() < st->get_priority()) {
+      this->stack_.insert(i, trigger);
+      this->current_status_ = "update me";
```
Comment (Unfocused) [inline esphome/components/status_indicator/status_indicator.cpp:164]: What's with "update me"? Temporary placeholder?

File: esphome/components/status_indicator/status_indicator.cpp [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+StatusTrigger *StatusIndicator::get_trigger(const std::string &key) {
+  auto search = this->triggers_.find(key);
+  if (search != this->triggers_.end()) {
+    return search->second;
+  } else {
+    return nullptr;
+  }
+}
+
+void StatusIndicator::set_trigger(const std::string &key, StatusTrigger *trigger) { this->triggers_[key] = trigger; }
```
Comment (Unfocused) [inline esphome/components/status_indicator/status_indicator.cpp:154]: It looks like it's not possible to have multiple handlers for a single status. That's not a problem when everything is in YAML, but I can imagine cases where C++ components would want to be able to handle those too.

Examples: the `status_led` component and light platform, possibly LVGL, plus third-party components (eg, EsphoMaTrix)

File: esphome/components/status_indicator/status_indicator.cpp [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+
+void StatusIndicator::dump_config() {
+  ESP_LOGCONFIG(TAG, "Status Indicator supports:");
+  for (auto i = this->triggers_.begin(); i != this->triggers_.end(); i++) {
+    ESP_LOGCONFIG(TAG, " * %s: %s", i->first.c_str(), i->second->get_info().c_str());
+  }
+}
+void StatusIndicator::loop() {
+  std::string status{""};
+  if ((App.get_app_state() & STATUS_LED_ERROR) != 0u) {
```
Comment (Unfocused) [inline esphome/components/status_indicator/status_indicator.cpp:65]: It feels awkward to have both pulling status (from core and some special components) and pushing status (custom statuses). I realise the pulling of info is coming from the existing way of doing things in the `status_led` component and light platform, but I think switching to/converging on just a push model makes the most sense for a few reasons.

* Custom statuses obviously work best with the push model
* A push model more easily allows additional data to be provided
* It would be the code that knows about that data (pushing it using a standard API), rather than having `status_indicator` try to reach into different core components using a different API for each status type

For an example of additional data: The code that handles marking components having errors/warnings (`Application::setup()` and `Application::loop()`) could report which components are having issues. At the moment, there's no way to tell - it's just a generic catch-all app error/warning. That's fine for a single LED or icon, but not useful for more detailed outputs.

As for using that additional data: It would be incredibly useful to show in a GUI on a screen (the dreaded "General error" versus the more useful "Error with temperature sensor" or "Wrong WiFi password"), or for custom logging (eg, logging to a service like Loki), etc.

TL;DR: It ends up feeling like this is still primarily designed for the blinking LED use-case (or a couple of icons), rather than something generic that can cover a wide variety of use-cases (including rich user interfaces).

File: esphome/components/status_indicator/status_indicator.cpp [hunk 5]
```diff
@@ -0,0 +1,213 @@
...
+#endif
+#ifdef USE_MQTT
+#include "esphome/components/mqtt/mqtt_client.h"
+#endif
+#ifdef USE_API
+#include "esphome/components/api/api_server.h"
+#endif
+
+namespace esphome {
+namespace status_indicator {
```
Comment (Unfocused) [inline esphome/components/status_indicator/status_indicator.cpp:19]: This would be the 3rd implementation for checking the status in ESPHome. The others are obviously the `status_led` component and `status_led` light platform - despite being in the same folder they duplicate code rather than share it. I do think it's important those continue to exist just because of how incredibly simple they are to use. But if/when this merges, I'd hope that they both rely this implementation rather than each having an alternative (divergent) implementations.

**Review by Unfocused — CHANGES_REQUESTED** [review]

**Comment by ptr727** [comment]
Thx, works great for boards with RGB LED's: https://github.com/ptr727/ESPHome-Config/blob/main/templates/rgb-led-status.yaml

**Comment by jwaz73** [comment]
+1 for this PR.  I like the ability to create a custom indicator as well.  Just got this working on an ESP32-S2 I'm playing with using the example from @ptr727 above.

**Comment by nielsnl68** [comment]
Oww, i did not saw there was so much interest in this PR. Sorry @Unfocused for not responding on your questions and suggestions. I will do that later this weekend.

After playing with it myself, i was not sure it did do as i was hoping it would do and i abended it until now.

**Comment by ptr727** [comment]
With the recent OTA and safemode changes, may I suggest to add a safe mode state?

**Comment by Totix92** [comment]
I am interested in this, I have two ESP32-S3 Supermini that have a rgb led, it would be nice if it was possible to use it with status_led

File: esphome/components/status_indicator/status_indicator.cpp (hunk 5 above)
Comment (nielsnl68) [reply esphome/components/status_indicator/status_indicator.cpp:19]: i agree on this.

File: esphome/components/status_indicator/status_indicator.cpp (hunk 4 above)
Comment (nielsnl68) [reply esphome/components/status_indicator/status_indicator.cpp:65]: Sorry for the late response, i love picking your brain a bit more about this. Would you mind pinging me on discord so we could have a chat?

File: esphome/components/status_indicator/status_indicator.cpp (hunk 1 above)
Comment (nielsnl68) [reply esphome/components/status_indicator/status_indicator.cpp:93]: true. When i started this PR i was not aware of this double use of the events.

File: esphome/components/status_indicator/status_indicator.cpp (hunk 3 above)
Comment (nielsnl68) [reply esphome/components/status_indicator/status_indicator.cpp:154]: I created this component just with one RGB led in mind. Your ideas are much versatile in mind. That is fine and they are reasonably ideas. As i wrote above lets have a chat about it see how we could extend this component in a later PR.

File: esphome/components/status_indicator/status_indicator.cpp (hunk 2 above)
Comment (nielsnl68) [reply esphome/components/status_indicator/status_indicator.cpp:164]: Yeah, i needed a way to notify the code below of the changes.

**Review by nielsnl68 — COMMENTED** [review]
Thanks @Unfocused for doing a review on this component.

File: esphome/components/status_indicator/__init__.py [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+    "on_app_warning": True,
+    "on_clear_app_warning": True,
+    "on_network_connected": True,
+    "on_network_disconnected": True,
+    "on_wifi_ap_enabled": True,
+    "on_wifi_ap_disabled": True,
+    "on_api_connected": True,
+    "on_api_disconnected": True,
+    "on_mqtt_connected": True,
+    "on_mgtt_disconnected": True,
```
Comment (mbronk) [inline esphome/components/status_indicator/__init__.py]:
```suggestion
    "on_mqtt_disconnected": True,
```
FYI, typo 😉

File: esphome/components/wifi/wifi_component.h [hunk 7]
```diff
@@ -264,6 +264,7 @@ class WiFiComponent : public Component {
 
   bool has_sta() const;
   bool has_ap() const;
+  bool is_ap_enabled() const;
```
Comment (mbronk) [inline esphome/components/wifi/wifi_component.h:270]:
```suggestion
  WiFiComponentState get_state() const { return this->state_; }
```
Just stumbled upon this PR accidentally. Great work! 🙏 

I'm no maintainer and not versed into this codebase so take with a fair grain of salt, but when looking through, I wondered if, _since you already extend `wifi` component_, it wouldn't be more futureproof to just expose the internal `state_` enum instead of yet another **bool** -> it'd allow other components to get more info about the WiFi state from one place.

I.e. I believe what you need could then still be accomplished via:
``` cpp
wifi::global_wifi_component->get_state() != wifi::WIFI_COMPONENT_STATE_AP
```

Comment (nielsnl68) [reply esphome/components/wifi/wifi_component.h:270]: When you modify the *StatusIndicator* component as well with your suggested change, i will commit this to this PR as well.

Anyway thanks for your contributions.

File: esphome/components/status_indicator/status_indicator.h [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+union StatusFlags {
+  struct {
+    int on_error : 1;
+    int on_warning : 1;
+    int on_network : 1;
+    int on_api : 1;
+    int on_mqtt : 1;
+    int on_wifi_ap : 1;
+  };
+  int setter = 0;
```
Comment (nielsnl68) [inline esphome/components/status_indicator/status_indicator.h]:
```suggestion
    uint16_t on_error : 1;
    uint16_t on_warning : 1;
    uint16_t on_network : 1;
    uint16_t on_api : 1;
    uint16_t on_mqtt : 1;
    uint16_t on_wifi_ap : 1;
  };
  uint16_t setter = 0;
```

File: esphome/components/wifi/wifi_component.h (hunk 7 above)
Comment (mbronk) [reply esphome/components/wifi/wifi_component.h:270]: Well.. I wasn't planning to upstream my tiny changes and merely wanted to hint this for your consideration, but... on a 2nd thought... why not 😁 Here you go: [GH-391](https://github.com/nielsnl68/esphome/pull/391).
Feel free to do with it as you please (edit/discard/merge...). I don't claim any authorship over this one-liner nor am I planning to extend it further.

Note it changes the semantics slightly: from `has AP enabled` to `has **only** AP enabled`. It is how that particular `status_` enum works internally and was more useful for me this way.

**Comment by AzonInc** [comment]
It seems like the custom status is not properly applied when rebooting and setting a status, maybe it's the boot priority or smth but I will try to give some instructions for reproduction.

**Comment by nielsnl68** [comment]
> It seems like the custom status is not properly applied when rebooting and setting a status, maybe it's the boot priority or smth but I will try to give some instructions for reproduction.
I'm not actively working on this PR. Could you give your YAML script and show where it is going wrong in your opinion, then i will look at it when i have time for it.

**Comment by Gunni** [comment]
I just tried to use this on my nanoESP32-C6-N16:

```yaml
external_components:
  - source: github://pr#5814
    components: [ status_indicator ]
```

But it fails compilation with:

```
src/esphome/components/status_indicator/status_indicator.cpp: In member function 'virtual void esphome::status_indicator::StatusIndicator::loop()':
src/esphome/components/status_indicator/status_indicator.cpp:75:56: error: 'class esphome::wifi::WiFiComponent' has no member named 'is_ap_enabled'; did you mean 'is_disabled'?
   75 |     if (status.empty() && wifi::global_wifi_component->is_ap_enabled()) {
      |                                                        ^~~~~~~~~~~~~
      |                                                        is_disabled
```

This was mentioned 8 months ago in a comment above, could you please fix it so we can use RGB Status LEDs?

File: esphome/components/wifi/wifi_component.h (hunk 7 above)
Comment (Gunni) [reply esphome/components/wifi/wifi_component.h:270]: I just tried to use this on my nanoESP32-C6-N16:

```yaml
external_components:
  #- source: github://pr#5814
  #  components: [ status_indicator ]
  - source: 
      type: git
      url: https://github.com/mbronk/esphome
      ref: private/mbronk/nvds-status-info-forked
    components: [ status_indicator ]
```

But it fails compilation with:

```
src/esphome/components/status_indicator/status_indicator.cpp: In member function 'virtual void esphome::status_indicator::StatusIndicator::loop()':
src/esphome/components/status_indicator/status_indicator.cpp:75:56: error: 'class esphome::wifi::WiFiComponent' has no member named 'get_state'; did you mean 'get_sta'?
   75 |     if (status.empty() && wifi::global_wifi_component->get_state() == wifi::WIFI_COMPONENT_STATE_AP) {
      |                                                        ^~~~~~~~~
      |                                                        get_sta
```

**Comment by nielsnl68** [comment]
> I just tried to use this on my nanoESP32-C6-N16:
> 
> ```yaml
> external_components:
>   - source: github://pr#5814
>     components: [ status_indicator ]
> ```

you also need to add "`wifi`" as component when you want to the *status_indicator* component.

**Comment by Gunni** [comment]
> you also need to add "`wifi`" as component when you want to the _status_indicator_ component.

```yaml
external_components:
  - source: github://pr#5814
    components: [ status_indicator, wifi ]
```

causes this:

```python
Traceback (most recent call last):
  File "/esphome/esphome/loader.py", line 200, in _lookup_module
    module = importlib.import_module(f"esphome.components.{domain}")
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/importlib/__init__.py", line 90, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<frozen importlib._bootstrap>", line 1387, in _gcd_import
  File "<frozen importlib._bootstrap>", line 1360, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1331, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 935, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 999, in exec_module
  File "<frozen importlib._bootstrap>", line 488, in _call_with_frames_removed
  File "/data/external_components/c0810ba7/esphome/components/wifi/__init__.py", line 4, in <module>
    from esphome.components.const import CONF_USE_PSRAM
ImportError: cannot import name 'CONF_USE_PSRAM' from 'esphome.components.const' (/esphome/esphome/components/const/__init__.py)
```

and this:

```yaml
external_components:
  #- source: github://pr#5814
  #  components: [ status_indicator, wifi ]
  - source:
      type: git
      url: https://github.com/mbronk/esphome
      ref: private/mbronk/nvds-status-info-forked
    components: [ status_indicator, wifi ]
```

causes:

```python
Traceback (most recent call last):
  File "/usr/local/bin/esphome", line 10, in <module>
    sys.exit(main())
             ^^^^^^
  File "/esphome/esphome/__main__.py", line 1087, in main
    return run_esphome(sys.argv)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/esphome/esphome/__main__.py", line 1065, in run_esphome
    config = read_config(dict(args.substitution) if args.substitution else {})
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/esphome/esphome/config.py", line 1162, in read_config
    res = load_config(command_line_substitutions)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/esphome/esphome/config.py", line 1016, in load_config
    return _load_config(command_line_substitutions)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/esphome/esphome/config.py", line 1006, in _load_config
    return validate_config(config, command_line_substitutions)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/esphome/esphome/config.py", line 902, in validate_config
    target_platform = core_config.preload_core_config(config, result)
                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/esphome/esphome/core/config.py", line 323, in preload_core_config
    if _is_target_platform(domain):
       ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/esphome/esphome/core/config.py", line 284, in _is_target_platform
    return get_component(name, True).is_target_platform
           ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/esphome/esphome/loader.py", line 224, in get_component
    return _lookup_module(domain, exception)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/esphome/esphome/loader.py", line 200, in _lookup_module
    module = importlib.import_module(f"esphome.components.{domain}")
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/importlib/__init__.py", line 90, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<frozen importlib._bootstrap>", line 1387, in _gcd_import
  File "<frozen importlib._bootstrap>", line 1360, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1331, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 935, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 999, in exec_module
  File "<frozen importlib._bootstrap>", line 488, in _call_with_frames_removed
  File "/data/external_components/577d1129/esphome/components/wifi/__init__.py", line 96, in <module>
    cv.Required(CONF_STATIC_IP): cv.ipv4,
                                 ^^^^^^^
AttributeError: module 'esphome.config_validation' has no attribute 'ipv4'
```

Having both just stops on the second problem...

**Comment by edwardtfn** [comment]
You should split the changes on wifi to another PR. It will be quite small and adding benefits by itself, so it will probably be approved much easier. And you don't have to merge `dev` all the time.
The it would be just about adding the new component...
